app/src/test/java/com/duckduckgo/app/sitepermissions/PermissionsPerWebsiteViewModelTest.kt

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2022 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.sitepermissions
+
+import app.cash.turbine.test
+import com.duckduckgo.app.CoroutineTestRule
+import com.duckduckgo.app.browser.R
+import com.duckduckgo.app.location.data.LocationPermissionEntity
+import com.duckduckgo.app.location.data.LocationPermissionType
+import com.duckduckgo.app.location.data.LocationPermissionsRepositoryAPI
+import com.duckduckgo.app.settings.db.SettingsDataStore
+import com.duckduckgo.app.sitepermissions.permissionsperwebsite.PermissionsPerWebsiteViewModel
+import com.duckduckgo.app.sitepermissions.permissionsperwebsite.PermissionsPerWebsiteViewModel.Command.GoBackToSitePermissions
+import com.duckduckgo.app.sitepermissions.permissionsperwebsite.PermissionsPerWebsiteViewModel.Command.ShowPermissionSettingSelectionDialog
+import com.duckduckgo.app.sitepermissions.permissionsperwebsite.WebsitePermissionSetting
+import com.duckduckgo.app.sitepermissions.permissionsperwebsite.WebsitePermissionSettingType
+import com.duckduckgo.app.statistics.pixels.Pixel
+import com.duckduckgo.site.permissions.impl.SitePermissionsRepository
+import com.duckduckgo.site.permissions.store.sitepermissions.SitePermissionAskSettingType
+import com.duckduckgo.site.permissions.store.sitepermissions.SitePermissionsEntity
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.runTest
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.stub
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.whenever
+
+@ExperimentalCoroutinesApi
+class PermissionsPerWebsiteViewModelTest {
+
+    @ExperimentalCoroutinesApi
+    @get:Rule
+    var coroutineRule = CoroutineTestRule()
+
+    private lateinit var viewModel: PermissionsPerWebsiteViewModel
+
+    private val mockSitePermissionsRepository: SitePermissionsRepository = mock()
+    private val mockLocationPermissionsRepository: LocationPermissionsRepositoryAPI = mock()
+    private val mockSettingsDataStore: SettingsDataStore = mock()
+    private val mockPixel: Pixel = mock()
+
+    private val domain = "domain.com"
+
+    @Before
+    fun before() {
+        viewModel = PermissionsPerWebsiteViewModel(
```

Comment:
NIT: we could move `viewmodel` as a class attribute, and move the other methods to the tests that need that setup. That way we won't need the `before()` method. This is optional because we don't have any established guidelines.

I have a personal preference to avoid, if possible, `before()` methods on tests. (Maybe I will propose it)
Basically, I prefer helper methods (the ones you already have like `loadAskForPermissionsPrefs`) and use them on each test case than having them in the `before()` method.
why?
- Easier to read tests if all required code is in each test case. Logic inside `before()` is usually hidden
- Avoid mocking objects that are not required on your tests
- if a test needs a different setup, you will need to override things that were initialized in the `before()`
- etc.

Reply:
I agree on this case because it's true that some of the mocks in `before()` are override in some functions. I will change it 👍